be/tool/beck.c

```diff
@@ -553,24 +588,81 @@ static char iobuf[4*1024*1024];
 
 enum { DELTA = 60 };
 
-static void generation_print(uint64_t gen)
+static void generation_id_print(uint64_t gen)
 {
 	time_t    ts = m0_time_seconds(gen);
 	struct tm tm;
 
 	localtime_r(&ts, &tm);
-	printf("%04d-%02d-%02d-%02d:%02d:%02d.%09lu",
+	printf("%04d-%02d-%02d-%02d:%02d:%02d.%09lu  (%lu)",
 	       tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
 	       tm.tm_hour, tm.tm_min, tm.tm_sec,
-	       m0_time_nanoseconds(gen));
+	       m0_time_nanoseconds(gen), gen);
+}
+
+static void generation_id_get(FILE *fp, uint64_t *gen_id)
+{
+	struct m0_format_tag    tag = {};
+	int                     type = M0_FORMAT_TYPE_BE_SEG_HDR;
+	struct m0_be_seg_hdr    seg_hdr;
+	const char             *rt_be_cksum;
+	int                     result = EX_OK;
+
+	if (fread(&seg_hdr, 1, sizeof(seg_hdr), fp) == sizeof(seg_hdr)) {
+
+		m0_format_header_unpack(&tag, &seg_hdr.bh_header);
+
+		if (seg_hdr.bh_header.hd_magic == M0_FORMAT_HEADER_MAGIC &&
+		    type == tag.ot_type &&
+		    memcmp(&tag, &rt[type].r_tag, sizeof tag) == 0 &&
+		    m0_format_footer_verify(&seg_hdr, 0) == 0) {
+
+			rt_be_cksum = m0_build_info_get()->
+						bi_xcode_protocol_be_checksum;
+			if (strncmp(seg_hdr.bh_be_version, rt_be_cksum,
+				    M0_BE_SEG_HDR_VERSION_LEN_MAX) == 0 &&
+			    seg_hdr.bh_items_nr == 1) {
+				/*
+				 * After confirming the segment header we can
+				 * read the embedded generation id.
+				 */
+				*gen_id = seg_hdr.bh_items[0].sg_gen;
+			} else
+				result = -EX_DATAERR;
```

Comment:
[* style *] `M0_ERR()`. The same in other relevant places.

What is there now is not useful. When someone will report 

> I have `-EX_DATAERR` error code, and I see "Invalid format / Checksum error for segment header" on the screen

what would you do? Go and ask something like "Please run custom build, then give us the output", or "run hexdump -C on <path to the file> and give us the output for offset 0x1234"?

General note: please save # of round-trips in case if something goes wrong.

Reply:
Done.